### DISCUSSION TITLE
Could not configure PAM for SSH authentication

### DISCUSSION THREAD

**Discussion by Tuharsky** [ltsp/ltsp#451] [Q&A]
Hallo, guys

I'm trying to run LTSP20 on Debian Bullseye (testing). I do it accordingly to https://ltsp.org/docs/installation/
I don't see any obvious problem during installation, it seems pretty easy, but the client bootup process ends with:

Could not configure PAM for SSH authentication!
LTSP command failed: /usr/share/ltsp/client/login/pamltsp install
Stopping plymouth
Aborting ltsp

Please, does anybody have an idea, what could be the culprit? Google dosen't give a clue.

**Comment by alkisg** [comment]
Hi Tuharsky,

this probably means that you have manually modified files under /etc/pam.d, which are normally managed by `pam-auth-update`.
First, keep a backup of /etc/pam.d.
Then try to run `pam-auth-update` on the server and see if it works or not.
If it doesn't, paste the errors.
If it does, can you upload your /etc/pam.d/common-auth ?

**Reply by Tuharsky** [reply]
Hi, Alkis
the pam-auth-update goes without error.

common-auth contains these active lines:
auth [success=1 default=ignore] pam_unix.so nullok
auth requisite pam_deny.so
auth required pam_permit.so

**Comment by alkisg** [comment]
Ah it might be an issue with your passwd files.
Can you paste the client boot errors BEFORE the ones you mentioned?
Ideally, upload a photo.

You may also try `sudo pwck` and `sudo grpck` on your server.

**Reply by Tuharsky** [reply]
Well, I have tried both, there were some problems with missing directories for lp, news, uucp, www-data, list, irc, gnats, nobody, _apt. This is fresh Debian installation, so there will be problems this way.
However, correcting these hasn't helped.
Here is photo:

![IMG20210426193719_00](https://user-images.githubusercontent.com/3784962/116130237-47511480-a6cb-11eb-965b-4181fddc8341.jpg)

**Comment by alkisg** [comment]
I just tried with bullseye-gnome and it worked fine here.
Put this in ltsp.conf, under [clients]:

```shell
PRE_INIT_DEBUG_PAMLTSP="
rs plymouth quit
sed '/^main /d' -i /usr/share/ltsp/client/login/pamltsp
cat >>/usr/share/ltsp/client/login/pamltsp <<EOF
set -x
exec 2>/tmp/pamltsp.$$
main "'\$@'
```

Then when it fails again, run `openvt bash -l` at the debug prompt, press Ctrl+Alt+F2 to switch to the shell at vt2, and from there, run `scp /tmp/pamltsp.1 user@server:` to copy it to your ltsp server, and then upload it here.
It will contain a shell trace of the pamltsp script.

**Reply by Tuharsky** [reply]
```
+ main install
+ local func _OLDIFS _SELF
+ _OLDIFS= 	

+ readlink -f /usr/share/ltsp/client/login/pamltsp
+ _SELF=/usr/share/ltsp/client/login/pamltsp
+ id -u
+ _UID=0
+ umask 0022
+ test -f /etc/ltsp/pamltsp.conf
+ SSH_SERVER=server
+ [ -z  ]
+ grep -qswE pam_sss|pam_ldap /etc/pam.d/common-auth
+ PAM_AUTH_TYPE=Primary
+ func=install
+ shift
+ install
+ local auth_control tab
+ require_root
+ [ 0 != 0 ]
+ [ Primary = Primary ]
+ auth_control=[success=end default=ignore]
+ printf Name: SSH based authentication for LTSP
Default: yes
Priority: 0
Auth-Type: %s
Auth:
\t%s\tpam_exec.so expose_authtok seteuid stdout quiet %s pam_auth
Session-Interactive-Only: yes
Session-Type: Additional
Session:
\toptional\tpam_exec.so seteuid stdout quiet %s pam_session\n Primary [success=end default=ignore] /usr/share/ltsp/client/login/pamltsp /usr/share/ltsp/client/login/pamltsp
+ pam-auth-update --package ltsp
Use of uninitialized value in concatenation (.) or string at /usr/share/perl5/Debconf/DbDriver/File.pm line 49, <DEBCONF_CONFIG> chunk 3.
Use of uninitialized value in -e at /usr/share/perl5/Debconf/DbDriver/File.pm line 51, <DEBCONF_CONFIG> chunk 3.
Use of uninitialized value in sysopen at /usr/share/perl5/Debconf/DbDriver/File.pm line 54, <DEBCONF_CONFIG> chunk 3.
Use of uninitialized value in concatenation (.) or string at /usr/share/perl5/Debconf/DbDriver/File.pm line 54, <DEBCONF_CONFIG> chunk 3.
debconf: DbDriver "config": could not open 
+ die Could not configure PAM for SSH authentication!
+ printf %s\n Could not configure PAM for SSH authentication!
Could not configure PAM for SSH authentication!
+ exit 1
```

**Comment by alkisg** [comment]
Sorry I had wrong escaping in the last line of the PRE_INIT_DEBUG_PAMLTSP directive.
I updated it in my previous comment, please see it.

**Reply by Tuharsky** [reply]
Thanks. Posted new output.

**Comment by alkisg** [comment]
Tuharsky you probably have a separate /var partition; you need to notify `ltsp image` about it otherwise it's missing from the generated image.
See: https://github.com/ltsp/ltsp/discussions/250#discussioncomment-247842

**Reply by Tuharsky** [reply]
Yes, that was it. Thank You.
Please, could it be mentioned in Wiki? I think, separate partitions may be quite common in the world of servers...

**Comment by alkisg** [comment]
It's in `man ltsp image`, see commit 85d647f.

**Reply by Tuharsky** [reply]
Well, I believe it's true, but the problem is, that it is not obvious that the inability to boot is connected to the image command, nor that it should be tackled around in any way. The wiki is pretty straightforward - run ltsp image / and whoila, everything is prepared.
It would suffice to put there a simple note, such as "If your system has divided partitions, see man page for the ltsp image command".

**Comment by alkisg** [comment]
Ah you mean in the https://ltsp.org/docs/installation/#maintaining-a-client-image web page (it's not a wiki).

You're probably right, I'll include it there as well, thank you.

**Reply by Tuharsky** [reply]
Well, You're right again, not a wiki (resembles one though)  :-)
I mean https://ltsp.org/docs/installation/
Thank You for Your patience.
